### DISCUSSION TITLE
Migration System for Pydantic Models

### DISCUSSION THREAD

**Discussion by unidesigner** [pydantic/pydantic#3685] [Question]
I was wondering if the following use case has already been discussed here or elsewhere, and whether there is some interest in the Pydantic community:

1. I have at one point in time defined a Pydantic model and generated instances of this model, e.g. in the form of JSON objects that are validated with the Pydantic model, and that I persist in a store such as in Postgres JSON column or a NoSQL database.
2. At a later time point, I need to make modifications to the original Pydantic model, thereby invalidating the previous JSON objects in the store
3. I want to automatically find the differences between the two Pydantic models (similar to e.g. a revision with the Alembic migration tool in SQL), i.e. generate the modifications/operations to go from the old model to the new model. 
4. I then would want to take this diff and transfer it into an executable form, perhaps modify it slightly, in order to run a migration to convert my old data into the new format which is then again validated with my new Pydantic model. The executable function could be just a Python function which takes a JSON and returns an updated JSON, or SQL statements for migrating etc.

Does anybody have any suggestions on how to move forward with this use case in a way that is as automated as possible, or perhaps there are existing projects along those lines?

**Comment by willyd** [comment]
I am also interested in this, but would not know where to start. The only project I could find that implements a migration of pydantic models is: [pydbantic](https://github.com/codemation/pydbantic#migrations).

**Comment by samuelcolvin** [comment]
I think this is really hard.

I also think in most cases, writing some custom logic to do the migration is pretty easy.

You might be able to auto-detect renamed fields, but it's not actually that common to rename fields.

Everything else would probably require custom logic.

**Reply by willyd** [reply]
@samuelcolvin I agree about the difficulty, but where would you suggest to add the custom logic? An easy solution would be to include a `schema_version: int` field in the model and use that to do the parsing/validation. My initial idea would be to use a `root_validator(pre=True)` or override `parse_obj` and check for the schema version in there. Would that make sense? For example, let's say I have `User` model intially defined as:
```
from typing import Any
from pydantic import BaseModel

class User(BaseModel):
    name: str
```
and then I change the model to:
```
from typing import Any
from pydantic import BaseModel

class User(BaseModel):
    schema_version: int = 2
    first_name: str
    last_name: str
```
Where would you place the logic to handle validation based on `schema_version`?

**Reply by samuelcolvin** [reply]
In a function somewhere else in the application, and call it just before initialising the model.

Or you could have a model which inherits from `NewUser` and add a root validator, e.g. `MigrationNewUser`.

I don't think this requires any changes to pydantic.

**Comment by zffocussss** [comment]
I am also curious about the migration tools/ways for the modification of models defination (like filed,type,etc.)
as Django has the migration script,It is more convenient if pydantic also have it

**Reply by zffocussss** [reply]
similar to @unidesigner case,the models data are store in my redis hashset.

**Reply by Viicos** [reply]
I suggest you take a look at the complexity of the [`django.db.migrations`](https://github.com/django/django/blame/main/django/db/migrations/) module before, as it involves complex graph logic. I also believe a migration system (at the model definition level)[^1] only makes sense if you are using some kind of DB backend (which is your case with Redis). In that case, it should probably be implemented in 3rd party libraries, e.g. https://github.com/redis/redis-om-python or https://github.com/art049/odmantic.

[^1]: Otherwise as said before manual editing of JSON instances can be implemented in a custom way

**Reply by zffocussss** [reply]
Hi @Viicos, I am investigating redis-om-python indeed .I am using redis for store user data (pydantic models).
